internal/spoke_k8s_client/factory.go

```diff
@@ -1,90 +1,255 @@
-	kubeconfig, err := kubeconfigFromSecret(secret)
+	// If the secret is from a hosted cluster then the API server address can be replaced by a
+	// 'kube-apiserver.*.svc' address that uses the service network and therefore reduces the number of round trips
+	// and doesn't need to go via proxies.
+	isHostedCluster, clusterDeploymentKey, err := f.isFromHostedCluster(ctx, secret)
 	if err != nil {
-		cf.log.WithError(err).Error("failed to get kubeconfig from secret")
-		return nil, nil, err
+		return
+	}
+	if isHostedCluster {
+		err = f.modifyRestConfigForHostedCluster(clusterDeploymentKey, restConfig)
+		if err != nil {
+			return
+		}
 	}
 
-	return cf.clientAndSetForKubeconfig(kubeconfig)
-}
+	// Add the transport wrappers:
+	for _, transportWrapper := range f.transportWrappers {
+		restConfig.Wrap(transportWrapper)
+	}
+
+	// Create the controller-runtime client:
+	client, err := f.clientFromRestConfig(restConfig)
+	if err != nil {
+		return
+	}
 
-func kubeconfigFromSecret(secret *corev1.Secret) ([]byte, error) {
-	if secret.Data == nil {
-		return nil, errors.Errorf("Secret %s/%s does not contain any data", secret.Namespace, secret.Name)
+	// Create the client-go client:
+	clientSet, err = f.clientSetFromRestConfig(restConfig)
+	if err != nil {
+		return
 	}
-	kubeconfigData, ok := secret.Data["kubeconfig"]
-	if !ok || len(kubeconfigData) == 0 {
-		return nil, errors.Errorf("Secret data for %s/%s does not contain kubeconfig", secret.Namespace, secret.Name)
+
+	// Create and populate the object:
+	result = &spokeK8sClient{
+		logger:      f.logger,
+		Client:      client,
+		csrClient:   clientSet.CertificatesV1().CertificateSigningRequests(),
+		sarClient:   clientSet.AuthorizationV1().SelfSubjectAccessReviews(),
+		nodesClient: clientSet.CoreV1().Nodes(),
 	}
-	return kubeconfigData, nil
+	return
 }
 
-func (cf *spokeK8sClientFactory) clientAndSetForKubeconfig(kubeconfig []byte) (SpokeK8sClient, *kubernetes.Clientset, error) {
-	clientConfig, err := clientcmd.NewClientConfigFromBytes(kubeconfig)
+func (f *spokeK8sClientFactory) restConfigFromSecret(secret *corev1.Secret) (result *rest.Config, err error) {
+	kubeConfig, err := f.kubeConfigFromSecret(secret)
 	if err != nil {
-		return nil, nil, errors.Wrapf(err, "failed to get clientconfig from kubeconfig data")
+		return
 	}
-
-	restConfig, err := clientConfig.ClientConfig()
+	clientConfig, err := clientcmd.NewClientConfigFromBytes(kubeConfig)
 	if err != nil {
-		return nil, nil, errors.Wrapf(err, "failed to get restconfig for kube client")
+		return
 	}
+	result, err = clientConfig.ClientConfig()
+	return
+}
 
-	clientset, err := kubernetes.NewForConfig(restConfig)
-	if err != nil {
-		cf.log.WithError(err).Warnf("Getting kuberenetes config for cluster")
-		return nil, nil, err
+func (f *spokeK8sClientFactory) kubeConfigFromSecret(secret *corev1.Secret) (result []byte, err error) {
```

Comment:
Can we make it common function? We have at least 2 more place that do the same

Reply:
Well, at least we have one less place now: I removed similar logic from the spoke client cache in a previous patch. I will try to find where we are doing this.

Reply:
I will do this in a different patch.